nemo/collections/nlp/data/datasets/sgd_dataset/evaluate.py

```diff
@@ -0,0 +1,209 @@
+PER_FRAME_OUTPUT_FILENAME = "dialogues_and_metrics.json"
+
+
+def get_service_set(schema_path):
+    """Get the set of all services present in a schema."""
+    service_set = set()
+    with open(schema_path) as f:
+        schema = json.load(f)
+        for service in schema:
+            service_set.add(service["service_name"])
+        f.close()
+    return service_set
+
+
+def get_in_domain_services(schema_path_1, schema_path_2):
+    """Get the set of common services between two schemas."""
+    return get_service_set(schema_path_1) & get_service_set(schema_path_2)
+
+
+def get_dataset_as_dict(file_path_patterns):
+    """Read the DSTC8 json dialog data as dictionary with dialog ID as keys."""
+    dataset_dict = {}
+    if isinstance(file_path_patterns, list):
+        list_fp = file_path_patterns
+    else:
+        list_fp = sorted(glob.glob(file_path_patterns))
+    for fp in list_fp:
+        if PER_FRAME_OUTPUT_FILENAME in fp:
+            continue
+        nemo.logging.info("Loading file: %s", fp)
+        with open(fp) as f:
+            data = json.load(f)
+            if isinstance(data, list):
+                for dial in data:
+                    dataset_dict[dial["dialogue_id"]] = dial
+            elif isinstance(data, dict):
+                dataset_dict.update(data)
+            f.close()
+    return dataset_dict
+
+
+def get_metrics(dataset_ref, dataset_hyp, service_schemas, in_domain_services, joint_acc_across_turn, no_fuzzy_match):
+    """Calculate the DSTC8 metrics.
+
+  Args:
+    dataset_ref: The ground truth dataset represented as a dict mapping dialogue
+      id to the corresponding dialogue.
+    dataset_hyp: The predictions in the same format as `dataset_ref`.
+    service_schemas: A dict mapping service name to the schema for the service.
+    in_domain_services: The set of services which are present in the training
+      set.
+    schemas: Schemas with information for all services
+
+  Returns:
+    A dict mapping a metric collection name to a dict containing the values
+    for various metrics. Each metric collection aggregates the metrics across
+    a specific set of frames in the dialogues.
+  """
+    # Metrics can be aggregated in various ways, eg over all dialogues, only for
+    # dialogues containing unseen services or for dialogues corresponding to a
+    # single service. This aggregation is done through metric_collections, which
+    # is a dict mapping a collection name to a dict, which maps a metric to a list
+    # of values for that metric. Each value in this list is the value taken by
+    # the metric on a frame.
+    metric_collections = collections.defaultdict(lambda: collections.defaultdict(list))
+
+    # Ensure the dialogs in dataset_hyp also occur in dataset_ref.
+    assert set(dataset_hyp.keys()).issubset(set(dataset_ref.keys()))
+    nemo.logging.info("len(dataset_hyp)=%d, len(dataset_ref)=%d", len(dataset_hyp), len(dataset_ref))
+
+    # Store metrics for every frame for debugging.
```

Comment:
hey, I am quite a lost in the following... is "frame"  a batch? what is happening here? are we iterating over the whole evaluation set in order to get the metrics?

Reply:
we're writing predictions to a file, then using the generated file and the original with labels to calculate all the metrics. SGD takes only current turn as an input but the dialogue state should carry the information about the previous steps as well.